### PR TITLE
WIP: Brave Rewards can now be completely reset.

### PR DESCRIPTION
**Pull request by jasonrsadler**
Fixes brave/brave-browser#1192

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests && npm run test-security`) on
  - [ ] Windows
  - [x] macOS
  - [ ] Linux
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:

1. Start Brave, clean profile and enable Rewards.
2. Perform actions on Rewards (contributions, Ads, etc)
3. Go to Settings->Advanced->Reset Rewards Settings
4. On the dialog click 'Reset Settings'
5. Ensure all files related to Rewards are removed from file system and that all Rewards pages are reverted to their 'pre-opt-in' state.
6. Enable Rewards again.
7. Ensure that a new wallet is created and that all Rewards functions perform correctly.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

**Comment by jasonrsadler** [comment]
cc @petemill - UI
@emerick - notifications
@tmancey - Ads/Confirmations
@NejcZdovc @ryanml - Rewards
@bbondy @bridiver - Chromium

**Comment by jasonrsadler** [comment]
Settings Bar
Screen Shot 2019-03-31 at 2 21 16 PM

Reset Brave Rewards Dialog
Screen Shot 2019-03-31 at 2 21 26 PM

Brave Dialog after reset attempt failed
Screen Shot 2019-04-02 at 11 42 33 AM

In the modal dialog above, the explanation text has updated to:
```
This will delete your Brave wallet and reset all of your Brave Rewards data, including your earned advertising tokens, contribution history, wallet transaction records, recurring tips to content creators, pending contributions to content creators, and any other Brave Rewards settings you may have saved.
```
the error text has been updated to:
```
Resetting Brave Rewards is failing at the moment. Please restart Brave and try again shortly.
```
Will update images shortly.
cc @mandar-brave

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 1]
```diff
@@ -2870,4 +2878,59 @@ void RewardsServiceImpl::OnDeleteActivityInfo(
...
+      file_task_runner_.get(), FROM_HERE,
+      base::BindOnce(&ResetClientOnFileTaskRunner, state_file_paths),
+      base::BindOnce(&RewardsServiceImpl::OnDeleteClientStateFiles,
+                     AsWeakPtr(), std::move(callback)));
+}
+
+void RewardsServiceImpl::OnDeleteClientStateFiles(
+    ledger::OnResetClientStateCallback callback,
+    bool success) {
+  publisher_info_backend_.reset(
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl.cc]: there is no reason to do this

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 2]
```diff
@@ -2870,4 +2878,59 @@ void RewardsServiceImpl::OnDeleteActivityInfo(
...
+      new PublisherInfoDatabase(publisher_info_db_path_));
+  notification_service_.reset(new RewardsNotificationServiceImpl(profile_));
+  auto* ads_service = brave_ads::AdsServiceFactory::GetForProfile(profile_);
+  if (ads_service)
+    ads_service->DisableAds();
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl.cc]: this doesn't seem like the right thing to do here. If we're resetting everything we need to reset ads too

File: patches/chrome-browser-resources-settings-route.js.patch [hunk 3]
```diff
@@ -27,11 +27,21 @@ index 337b8c13d19ca4ed3848057608f4a6de62dde82e..14dd9a6c8b348aedb0e482050d8c0c26
...
        r.CLOUD_PRINTERS = r.PRINTING.createChild('/cloudPrinters');
 +      // </if>
        // <if expr="chromeos">
        r.CUPS_PRINTERS = r.PRINTING.createChild('/cupsPrinters');
        // </if>
+@@ -419,6 +427,9 @@ cr.define('settings', function() {
+         r.TRIGGERED_RESET_DIALOG =
+             r.ADVANCED.createChild('/triggeredResetProfileSettings');
+         r.TRIGGERED_RESET_DIALOG.isNavigableDialog = true;
++        r.RESET_REWARDS_DIALOG =
```
Comment (bridiver) [inline patches/chrome-browser-resources-settings-route.js.patch]: there is now a better way to do this stuff cc @petemill

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 4]
```diff
@@ -2870,4 +2878,59 @@ void RewardsServiceImpl::OnDeleteActivityInfo(
...
+  std::move(callback).Run(success);
+  for (auto& observer_ : observers_) {
+    observer_.OnReset(this);
+  }
+}
+
+void RewardsServiceImpl::DeleteClientStateFiles(
+    ledger::OnResetClientStateCallback callback) {
+#if defined(OS_WIN)
+  const base::FilePath::StringType
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl.cc]: we should be clearing the database, not deleting the db files

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 5]
```diff
@@ -2870,4 +2878,59 @@ void RewardsServiceImpl::OnDeleteActivityInfo(
...
+      base::BindOnce(&RewardsServiceImpl::OnDeleteClientStateFiles,
+                     AsWeakPtr(), std::move(callback)));
+}
+
+void RewardsServiceImpl::OnDeleteClientStateFiles(
+    ledger::OnResetClientStateCallback callback,
+    bool success) {
+  publisher_info_backend_.reset(
+      new PublisherInfoDatabase(publisher_info_db_path_));
+  notification_service_.reset(new RewardsNotificationServiceImpl(profile_));
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl.cc]: this also doesn't seem right. We're clearing the notifications so why would we create a  new RewardsNotificationServiceImpl?

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 6]
```diff
@@ -2870,4 +2878,59 @@ void RewardsServiceImpl::OnDeleteActivityInfo(
...
+#if defined(OS_WIN)
+  const base::FilePath::StringType
+      kPublisher_info_db_journal(L"publisher_info_db-journal");
+#else
+  const base::FilePath::StringType
+      kPublisher_info_db_journal("publisher_info_db-journal");
+#endif
+  const base::FilePath publisher_info_db_journal_path_(
+      profile_->GetPath().Append(kPublisher_info_db_journal));
+  publisher_info_backend_.release();
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl.cc]: these `release` calls leak memory

File: components/brave_ads/browser/ads_service.h [hunk 7]
```diff
@@ -41,6 +41,7 @@ class AdsService : public KeyedService {
   virtual void SetConfirmationsIsReady(const bool is_ready) = 0;
   virtual void IsSupportedRegion(
       IsSupportedRegionCallback callback) = 0;
+  virtual void DisableAds() = 0;
```
Comment (bridiver) [inline components/brave_ads/browser/ads_service.h]: there is already a method to enable/disable ads, why are we adding a new one?

File: components/brave_ads/browser/ads_service_impl.cc [hunk 8]
```diff
@@ -311,6 +311,11 @@ void AdsServiceImpl::OnCreate() {
   }
 }
 
+void AdsServiceImpl::DisableAds() {
+  set_ads_enabled(false);
```
Comment (bridiver) [inline components/brave_ads/browser/ads_service_impl.cc]: calling set_ads_enabled(false) already calls Stop

File: patches/chrome-browser-resources-settings-reset_page-reset_page.html.patch [hunk 9]
```diff
@@ -0,0 +1,21 @@
+diff --git a/chrome/browser/resources/settings/reset_page/reset_page.html b/chrome/browser/resources/settings/reset_page/reset_page.html
+index 2910641f45437a6461433f1fde8db6c36f7780db..ffd7056e8c37d7804de0813272e578f90e7ca33f 100644
+--- a/chrome/browser/resources/settings/reset_page/reset_page.html
++++ b/chrome/browser/resources/settings/reset_page/reset_page.html
+@@ -10,6 +10,8 @@
+ <link rel="import" href="../settings_page/settings_animated_pages.html">
+ <link rel="import" href="../settings_shared_css.html">
+ 
++<link rel="import" href="../brave_reset_page/brave_reset_page.html">
```
Comment (bridiver) [inline patches/chrome-browser-resources-settings-reset_page-reset_page.html.patch]: cc @petemill

File: components/services/bat_ledger/public/interfaces/bat_ledger.mojom [hunk 10]
```diff
@@ -179,4 +180,5 @@ interface BatLedgerClient {
   ConfirmationsTransactionHistoryDidChange();
 
   GetExcludedPublishersNumberDB() => (uint32 number);
+  DeleteClientStateFiles() => (bool success);
```
Comment (bridiver) [inline components/services/bat_ledger/public/interfaces/bat_ledger.mojom]: The LedgerClient shouldn't have to know what needs to happen to delete the client state. That is likely to lead to inconsistent implementations. We already have `ResetState` methods for the different state files and I am adding delete methods for the DB

File: components/brave_rewards/browser/rewards_service_impl.cc (hunk 4 above)
Comment (jasonrsadler) [reply components/brave_rewards/browser/rewards_service_impl.cc]: There are is no db file pre-opt-in

File: components/services/bat_ledger/public/interfaces/bat_ledger.mojom [hunk 11]
```diff
@@ -103,6 +103,7 @@ interface BatLedger {
   GetConfirmationsHistory(uint64 from_timestamp_seconds,
       uint64 to_timestamp_seconds) => (string transactions);
   GetRewardsInternalsInfo() => (string info);
+  DeleteStateFiles() => (bool success);
```
Comment (bridiver) [inline components/services/bat_ledger/public/interfaces/bat_ledger.mojom]: These will not always be "Files". What we are asking Ledger to do is to reset the state so I think this method should be `ResetState()`. Also we shouldn't be using `bool` for success/failure, that's what ledger::Result is for

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 12]
```diff
@@ -2870,4 +2878,59 @@ void RewardsServiceImpl::OnDeleteActivityInfo(
...
+void RewardsServiceImpl::OnReset(
+    ResetStateCallback callback,
+    bool success) {
+  std::move(callback).Run(success);
+  for (auto& observer_ : observers_) {
+    observer_.OnReset(this);
+  }
+}
+
+void RewardsServiceImpl::DeleteClientStateFiles(
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl.cc]: as mentioned in the other comment, we can't put the responsibility for knowing what needs to be cleared/deleted on the LedgerClient. The LedgerClient methods are just to perform IO and other operations that the Ledger library is not allowed to perform. We already have save/load/reset state methods and that is what should be used for clearing the state.

File: components/services/bat_ledger/public/interfaces/bat_ledger.mojom (hunk 11 above)
Comment (jasonrsadler) [reply components/services/bat_ledger/public/interfaces/bat_ledger.mojom]: I'm open to any naming convention suggestions as `ResetState()` seems to be used in quite a few places.

Also the end result determines whether or not to show the error so ultimately we would be return a bool `(result == LEDGER_OK)`

File: components/brave_rewards/browser/rewards_service_impl.cc [hunk 13]
```diff
@@ -2870,4 +2878,59 @@ void RewardsServiceImpl::OnDeleteActivityInfo(
...
+void RewardsServiceImpl::OnDeleteClientStateFiles(
+    ledger::OnResetClientStateCallback callback,
+    bool success) {
+  publisher_info_backend_.reset(
+      new PublisherInfoDatabase(publisher_info_db_path_));
+  notification_service_.reset(new RewardsNotificationServiceImpl(profile_));
+  auto* ads_service = brave_ads::AdsServiceFactory::GetForProfile(profile_);
+  if (ads_service)
+    ads_service->DisableAds();
+  callback(success);
```
Comment (bridiver) [inline components/brave_rewards/browser/rewards_service_impl.cc]: we're resetting all these things that don't need to be reset, but we're not resetting the one thing that does need to be reset which is the bat_ledger service

**Comment by bridiver** [comment]
this definitely needs tests

File: components/brave_rewards/browser/rewards_service_impl.cc (hunk 4 above)
Comment (bridiver) [reply components/brave_rewards/browser/rewards_service_impl.cc]: That doesn't make any difference. First of all it's not safe or reliable to try to delete the db files and second the existence of the files is irrelevant. What matters is whether there is data in them or not.

File: components/services/bat_ledger/public/interfaces/bat_ledger.mojom (hunk 11 above)
Comment (bridiver) [reply components/services/bat_ledger/public/interfaces/bat_ledger.mojom]: the reason ledger::Result exists is to provide better information about failures. We should want consistency and that's why we use ledger::Result everywhere and shouldn't be using `bool`. if `result != LEDGER_OK` then how do you know what happened or how to attempt to resolve the problem if all you get is `false`?

Comment (bridiver) [reply components/services/bat_ledger/public/interfaces/bat_ledger.mojom]: also `ResetState` is used in a few places and that's exactly what they are there for. We call `ResetState` on Ledger which in turn should be calling `ResetState` for everything it would normally call `LoadState`/'SaveState` on and then also calling db table delete methods

File: vendor/bat-native-confirmations/src/bat/confirmations/internal/confirmations_impl.cc [hunk 14]
```diff
@@ -1055,4 +1055,22 @@ bool ConfirmationsImpl::IsRetryingToGetRefillSignedTokens() const {
...
+  state_has_loaded_ = false;
+  confirmations_client_->ResetState(_confirmations_name,
+      std::bind(&ConfirmationsImpl::OnResetConfirmationsState,
+        this, callback, _1));
+}
+
+void ConfirmationsImpl::OnResetConfirmationsState(
+    OnResetConfirmationsStateCallback callback,
+    const ledger::Result result) {
+  StopRetryingToGetRefillSignedTokens();
```
Comment (bridiver) [inline vendor/bat-native-confirmations/src/bat/confirmations/internal/confirmations_impl.cc]: this seems very fragile. We need a more generalized Shutdown method for all of Ledger to stop anything currently in progress and that should be called before deleting the state

File: components/brave_ads/browser/ads_service.h (hunk 7 above)
Comment (tmancey) [reply components/brave_ads/browser/ads_service.h]: We should be disabling Ads from `bat-native-ads` lib

File: vendor/bat-native-ledger/include/bat/ledger/ledger.h [hunk 15]
```diff
@@ -62,6 +62,8 @@ using ConfirmationsHistoryCallback = std::function<void(
 using GetExcludedPublishersNumberDBCallback = std::function<void(uint32_t)>;
 using OnWalletPropertiesCallback = std::function<void(const ledger::Result,
                                   std::unique_ptr<ledger::WalletInfo>)>;
+using OnResetStateCallback = std::function<void(bool)>;
+using ResetConfirmationsStateCallback = std::function<void(bool)>;
```
Comment (bridiver) [inline vendor/bat-native-ledger/include/bat/ledger/ledger.h]: why is this here?

File: vendor/bat-native-confirmations/src/bat/confirmations/internal/confirmations_impl.cc [hunk 16]
```diff
@@ -1055,4 +1055,22 @@ bool ConfirmationsImpl::IsRetryingToGetRefillSignedTokens() const {
   return true;
 }
 
+void ConfirmationsImpl::ResetConfirmationsState(
+    OnResetConfirmationsStateCallback callback) {
+  is_initialized_ = false;
+  state_has_loaded_ = false;
+  confirmations_client_->ResetState(_confirmations_name,
```
Comment (bridiver) [inline vendor/bat-native-confirmations/src/bat/confirmations/internal/confirmations_impl.cc]: this is what LedgerImpl should be doing - calling ResetState for the relevant state files

File: vendor/bat-native-ledger/include/bat/ledger/ledger.h (hunk 15 above)
Comment (jasonrsadler) [reply vendor/bat-native-ledger/include/bat/ledger/ledger.h]: https://github.com/brave/brave-core/pull/2141/files#diff-a293ba328fc758a3ff87a1d08a02329aR36

File: components/brave_rewards/browser/rewards_service_impl.cc (hunk 6 above)
Comment (bridiver) [reply components/brave_rewards/browser/rewards_service_impl.cc]: this also isn't allowed on the UI thread for `publisher_info_backend_`

File: vendor/bat-native-ledger/include/bat/ledger/ledger.h (hunk 15 above)
Comment (bridiver) [reply vendor/bat-native-ledger/include/bat/ledger/ledger.h]: there is already a callback defined for the LedgerClient::ResetState method which uses the current ledger::Result return type (using OnResetCallback = std::function<void(const ledger::Result)>;), but either way, why would you put it here? It isn't used in this file and isn't relevant at all to ledger.h

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc [hunk 17]
```diff
@@ -1364,6 +1364,25 @@ void LedgerImpl::GetConfirmationsHistory(
...
+    ledger::OnResetStateCallback callback) {
+  ResetState();
+  bat_confirmations_->ResetConfirmationsState(
+      std::bind(&LedgerImpl::OnResetConfirmationsState,
+        this, callback, _1));
+}
+
+void LedgerImpl::OnResetConfirmationsState(
+    ledger::OnResetStateCallback callback,
+    bool confirmations_result) {
```
Comment (bridiver) [inline vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: result is ignored?

File: patches/chrome-browser-resources-settings-reset_page-reset_page.html.patch (hunk 9 above)
Comment (mkarolin) [reply patches/chrome-browser-resources-settings-reset_page-reset_page.html.patch]: This is still pending review: https://github.com/brave/brave-core/pull/2130

File: components/services/bat_ledger/public/interfaces/bat_ledger.mojom (hunk 11 above)
Comment (jasonrsadler) [reply components/services/bat_ledger/public/interfaces/bat_ledger.mojom]: `result != LEDGER_OK` is to report to the UI as a general failure to the user with "Try later" or "Try restarting and try again...". Sure, we can report back `ledger::DB_LOCKED` or something similar, but the resulting explanation to the user would be relatively the same.

File: patches/chrome-browser-resources-settings-route.js.patch (hunk 3 above)
Comment (jasonrsadler) [reply patches/chrome-browser-resources-settings-route.js.patch]: Is there documentation that you can link me to?

File: components/brave_rewards/browser/rewards_service_impl.cc (hunk 4 above)
Comment (jasonrsadler) [reply components/brave_rewards/browser/rewards_service_impl.cc]: Would the state files (`ledger_state`, `publisher_state`, `confirmations.json`, etc) remain as well or be deleted?

File: components/brave_rewards/browser/rewards_service_impl.cc (hunk 13 above)
Comment (jasonrsadler) [reply components/brave_rewards/browser/rewards_service_impl.cc]: So `bat_ledger_` is the only things that should get reset?

File: vendor/bat-native-confirmations/src/bat/confirmations/internal/confirmations_impl.cc (hunk 14 above)
Comment (jasonrsadler) [reply vendor/bat-native-confirmations/src/bat/confirmations/internal/confirmations_impl.cc]: cc @tmancey is there a better way?

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc (hunk 17 above)
Comment (jasonrsadler) [reply vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc]: Thanks for catching. Will handle.

File: components/brave_ads/browser/ads_service.h (hunk 7 above)
Comment (jasonrsadler) [reply components/brave_ads/browser/ads_service.h]: I had created this as I missed the PrefChanged event on `enable_ads(bool)`. Will fix.

File: components/brave_ads/browser/ads_service_impl.cc (hunk 8 above)
Comment (jasonrsadler) [reply components/brave_ads/browser/ads_service_impl.cc]: Yep. Missed the PrefChanged event.

**Comment by bsclifton** [comment]
Closing as this is a bit stale and has been de-prioritized
